### DISCUSSION TITLE
Customization excludes

### DISCUSSION THREAD

**Discussion by MichaelDvP** [emsesp/EMS-ESP32#405] [Ideas]
The numbering, changing on every new value, is getting a bit nasty. How about using the short_names instead of numbers. The file will grow a bit, but only ~2k max. And there is no limit to 255 entities.

This is the result of a test:
[emsesp_customizations.txt](https://github.com/emsesp/EMS-ESP32/files/8309349/emsesp_customizations.txt)

**Comment by proddy** [comment]
Agree, it's getting annoying. I broke my HA system too (how's that for eating your own dog food ;-))

I've been thinking how best to do #317 and using a flag byte (b01 = excluded from MQTT, b10 = make it read-only) and adding to the JSON array (using an key/value pair would take too much memory). For example:

```
    "exclude_entities": [
      {
        "product_id": 72,
        "device_id": 8,
        "entity_ids": [
          "burngas2", 1,
          "pumpmodmax", 2,
          "pumpmodmin", 3,
          "burnminpower", 1
        ]
      }

```

the exclude_entities would need renamed to something more meaningful like 'custom_entities', which would be a breaking change.

**Reply by proddy** [reply]
and perhaps we should do this in 3.4 while we can so it doesn't break future versions.

**Reply by MichaelDvP** [reply]
Can you make an array with string/number mixing? You can also add a first (or last) characters as a hex 00..FF, making the entity to `03hc2/tempautotemp`, this saves a bit storage space.
There was also #344, what flags are you planing? `exclude mqtt`, `exclude from web`, `exclude from api`, `read-only`, `favorite`

**Reply by proddy** [reply]
great idea to prefix the shortname with the hex flags. Saves a lot of space.

For flags, haven't really thought about it but yes something like
 - exclude api/mqtt
 - exclude web
 - force read-only
 - favorite
 
I think the console should show everything as it's only used for power users.

I think we should make the change now to the JSON, release v3.4 and add the backend code and Web page to edit the flags later?

**Comment by proddy** [comment]
@MichaelDvP do we need to show hidden entities in https://github.com/emsesp/EMS-ESP32/blob/28ae1ae3d4a07a8bc162ecf08b4ba0825c2042ff/src/emsdevice.cpp#L836 ?

it currently shows up as:

![image](https://user-images.githubusercontent.com/1230712/160351462-378efee3-aaac-41af-9975-c09859f5005a.png)

I say just filter out anything that `dv.full_name` is !null

**Reply by MichaelDvP** [reply]
It was intended to show `climate` on the exclusion screen to allow disable from ha-config creation. But the value/name is always hidden.

**Reply by proddy** [reply]
ok, so keep it

**Reply by MichaelDvP** [reply]
We can replace the `(hidden)` by an empty string and force the web-exclude instead.  I think this is more clear:
![grafik](https://user-images.githubusercontent.com/59284019/160447998-719de0d4-323b-4510-801c-5dfe1f6f2ed0.png)
```
    const getMask = (de: DeviceEntity) => {
      var new_masks = [];
      if ((de.m & 1) === 1 || de.n === '') {
        new_masks.push('1');
      }
```

**Reply by proddy** [reply]
ok, you add or me?

**Comment by proddy** [comment]
one thing I noticed is that when selected DV_READONLY on one of the Entities it won't automatically delete the old HA config topic and re-create it. I tried modifying the code but it's tricky to catch because we don't recognize state changes so you're not sure which HA topic to remove. So solution is just to restart EMS-ESP after any entities have been marked as read-only and using Home Assistant.

see https://github.com/emsesp/EMS-ESP32/issues/317#issuecomment-1082678466

**Reply by MichaelDvP** [reply]
We need an extra flag for the HA readonly-status and sync the flags after each publish. Re-publish with new state should do, a remove is not needed. I'll make aPR.

**Reply by proddy** [reply]
I was thinking it's easier to just change EMSdevice::mask_entity() which is called only after a Save click

and in that function look for entities that have the DV_READONLY flag recently changed and then remove it with the code:
```
dv.remove_state(DeviceValueState::DV_HA_CONFIG_CREATED);
Mqtt::publish_ha_sensor_config(dv, "", "", true); // delete topic (remove = true)
```

no need to add more code in EMSdevice::mqtt_ha_entity_config_create() which is costly as it's called every few seconds.

**Reply by MichaelDvP** [reply]
Agree, that's the better solution, change detect in a short form: `if ((dv.state ^ (flag << 4)) & DeviceValueState::DV_READONLY) {`

**Comment by proddy** [comment]
@MichaelDvP bbqkees have some feedback this morning, he said should 'reset' be default hidden from the WebUI ?

**Reply by MichaelDvP** [reply]
Why? There was a long discussion how to add the boiler reset key to ems-esp. 
Afair there is a little risk if someone uses the reset too often in a row. Then the boiler goes to blocking error and needs a reset on the burner or a powercycle. But it's like pushing the reset key on the boiler, this key is not hidden and can be pressed by anyone. It's only resetting error or maintenance message, not changing any setting.

**Reply by proddy** [reply]
@bbqkees ?

**Comment by bbqkees** [comment]
Ok I thought some more settings would be reset if it is pressed.
if I check the manual of my boiler it's more like a reboot button. It only clears the error and reboots.

In that case you may want to rename it to 'reset error' or something.

**Reply by MichaelDvP** [reply]
It has two functions, reset error messages and reset maintenance messages, this can be selected. The default empty selection will not send any ems command, also command via mqtt/api needs a parameter to be executed. On my boiler it only clears the error, not rebooting. If there is no error, it do nothing (checked a minute ago).

On the boiler the button is named "Reset", i don't think anything is more clear if we use a different name.
![grafik](https://user-images.githubusercontent.com/59284019/160831020-06fdb2f5-163e-48c3-936d-c594b6c9e669.png)
